Add a filter to hide/disable the admin controls

bernattorras opened the pull request:
### Description

As detailed in #43, it would be nice to be able to hide the Safety Net admin page (Tools > Safety Net) using a filter. This way, any developer (or us) could disable it easily by using the following code:

`add_filter( 'safety_net_hide_admin', '__return_true' );`

bernattorras committed:
Moved all the hooks into a function

bernattorras committed:
Add a filter to hide the Safety Net options page

Any developer could disable the `Safety Net` admin page (Tools > Safety Net) simply by using this filter.

Example:

`add_filter( 'safety_net_hide_admin', '__return_true' );`

NickGreen commented:
Most of the functions that you've allowed to be filtered out there are functions that we actually want to run, regardless of if the admin controls are showing or not. For example, `add_filter( 'wp_mail', __NAMESPACE__ . '\stop_emails', 10, 1 );` is the function that stops emails. 

I think the admin tools that he meant are the actual buttons and toggle at **Tools > Safety Net**, as seen here:
![Screen Shot on 2022-11-17 at 14:29:39](https://user-images.githubusercontent.com/2067992/202573910-6164c153-1585-424b-aaa3-4369517fb872.png)

So what you'll probably want to do is only include the hooks in that filterable function which render the wp-admin menu item, the link on the plugins page, and the one that renders the Safety Net Tools page.

bernattorras commented:
Thanks for the head-ups @NickGreen! Yeah, I missed these filters! 

I'm a bit doubtful though about just hiding the controls... Would it make sense to have the Safety Net page available in the admin if you can't perform any actions from there? What would be the benefit? To check the possible features that aren't available for the user? 

<img width="1526" alt="Markup 2022-11-22 at 13 25 07" src="https://user-images.githubusercontent.com/1201868/203313450-cd01dfa5-1922-47f0-a70d-e4734eed7506.png">

I think that since we already display the admin notice in the backend (and pause the emails, scheduled actions, etc), we could directly hide the entire Safety Net options page, but that's just my opinion. :) 

What do you think @nate-allen @NickGreen? Could you share your thoughts and possible cases to find out the best approach?

NickGreen commented:
Yeah, I agree, it would be fine to hide the whole page and the links to the page (there's one in the **Tools** menu and one on the plugin page).

The main thing is that we don't want to stop _all_ of the functions that load in admin, just the ones that render the page and menu items.

nate-allen commented on includes/admin.php:
```diff
@@ -8,17 +8,31 @@
+	add_action( 'action_scheduler_pre_init', __NAMESPACE__ . '\pause_renewal_actions' );
+	add_action( 'admin_notices', __NAMESPACE__ . '\show_warning' );
+	add_filter( 'plugin_action_links_' . SAFETY_NET_BASENAME, __NAMESPACE__ . '\add_action_links' );
+	add_filter( 'wp_mail', __NAMESPACE__ . '\stop_emails', 10, 1 );
```
We probably would still want to block emails, right?

nate-allen commented on includes/admin.php:
```diff
@@ -8,17 +8,31 @@
+	add_action( 'wp_ajax_safety_net_deactivate_plugins', __NAMESPACE__ . '\handle_ajax_deactivate_plugins' );
+	add_action( 'wp_ajax_safety_net_delete_users', __NAMESPACE__ . '\handle_ajax_delete_users' );
+	add_action( 'action_scheduler_pre_init', __NAMESPACE__ . '\pause_renewal_actions' );
+	add_action( 'admin_notices', __NAMESPACE__ . '\show_warning' );
```
We should still show the warning that Safety Net is enabled

nate-allen commented on includes/admin.php:
```diff
@@ -8,17 +8,31 @@
+	add_action( 'wp_ajax_safety_net_scrub_options', __NAMESPACE__ . '\handle_ajax_scrub_options' );
+	add_action( 'wp_ajax_safety_net_deactivate_plugins', __NAMESPACE__ . '\handle_ajax_deactivate_plugins' );
+	add_action( 'wp_ajax_safety_net_delete_users', __NAMESPACE__ . '\handle_ajax_delete_users' );
+	add_action( 'action_scheduler_pre_init', __NAMESPACE__ . '\pause_renewal_actions' );
```
I think we would still want to pause renewals

nate-allen reviewed (dismissed):
I left some comments. It looks like to me that there are still some things being blocked that we would want to be active even if the admin controls are hidden.

bernattorras committed:
Skip only the admin options related hooks.

Skip the admin page and options if the `safety_net_hide_admin` filter returns true.
The other hooks ('pause_renewal_actions', 'show_warning' and 'stop_emails' are still required)

bernattorras commented:
Thanks @nate-allen! Yeah, these are the ones that I had in my mind, but I just wanted to clarify first if we still needed to load the options page (without the controls) or hide it completely. 👍

nate-allen commented:
We should hide it completely, even removing "Safety Net" link from the Tools menu.

bernattorras committed:
Merge branch 'trunk' into feature/issue_43
